src/main/java/net/dv8tion/jda/core/utils/PermissionUtil.java

```diff
@@ -420,79 +432,179 @@ public static long getEffectivePermission(Channel channel, Role role)
+     *
+     * @return Primitive (unsigned) long value with the implicit permissions of the specified role in the specified channel
+     *
+     * @since  3.1
+     */
+    public static long getImplicitPermission(Channel channel, Role role)
     {
-        long permission = getEffectivePermission(guild, member);
+        Args.notNull(channel, "Channel");
+        Args.notNull(role, "Role");
 
-        //override with channel-specific overrides of @everyone
-        PermissionOverride override = roleOverrides.get(guild.getPublicRole());
+        final Guild guild = role.getGuild();
+        checkGuild(channel.getGuild(), guild, "Role");
+
+        long permission = role.getPermissionsRaw() & guild.getPublicRole().getPermissionsRaw();
+        PermissionOverride override = channel.getPermissionOverride(guild.getPublicRole());
         if (override != null)
-        {
             permission = apply(permission, override.getAllowedRaw(), override.getDeniedRaw());
+        if (role.equals(guild.getPublicRole()))
+            return permission;
+
+        override = channel.getPermissionOverride(role);
+
+        return override == null
+            ? permission
+            : apply(permission, override.getAllowedRaw(), override.getDeniedRaw());
+    }
+
+    private static void getImplicitOverrides(Channel channel, Member member, AtomicLong allow, AtomicLong deny)
+    {
+        PermissionOverride override = channel.getPermissionOverride(member.getGuild().getPublicRole());
+        if (override != null)
+        {
+            deny.set(override.getDeniedRaw());
+            allow.set(override.getAllowedRaw());
         }
 
-        //handle role-overrides of this member in this channel (allow > disallow)
-        long allow = -1;
-        long deny = -1;
         for (Role role : member.getRoles())
         {
-            PermissionOverride po = roleOverrides.get(role);
-            if (po != null) //If an override exists for this role
+            override = channel.getPermissionOverride(role);
+            if (override != null)
             {
-                if (allow == -1 || deny == -1)  //If this is the first role we've encountered.
-                {
-                    allow = po.getAllowedRaw(); //First role, take values from this role as the base for permission
-                    deny = po.getDeniedRaw();   //
-                }
-                else
-                {
-                    allow = po.getAllowedRaw() | allow;     //Give all the stuff allowed by this Role's allow
-
-                    deny = (po.getDeniedRaw() | deny) & (~allow);  //Deny everything that this role denies.
-                    // This also rewrites the previous role's denies if this role allowed those permissions.
-                }
+                deny.accumulateAndGet(override.getDeniedRaw(), PermissionUtil::accumulate);
+                allow.accumulateAndGet(override.getAllowedRaw(), PermissionUtil::accumulate);
```

Comment:
no, I meant the `PermissionUtil::accumulate`

`allow.accumulateAndGet((val1, val2) -> val1 | val2);`